src/Microsoft.SqlTools.Migration/MigrationService.cs

```diff
@@ -320,6 +324,138 @@ public void InitializeService(IProtocolEndpoint serviceHost)
             }
         }
 
+        internal async Task HandleSysAdminPermissionValidation(
+            StartLoginMigrationParams parameters,
+            RequestContext<LoginMigrationPreValidationResult> requestContext)
+        {
+            try
+            {
+                ILoginsMigrationLogger logger = this.GetLoginsMigrationLogger();
+                ILoginsMigration loginMigration = new LoginsMigration(parameters.SourceConnectionString, parameters.TargetConnectionString,
+                null, parameters.LoginList, parameters.AADDomainName, logger);
+
+                IDictionary<string, IEnumerable<LoginMigrationException>> exceptionMap = new Dictionary<string, IEnumerable<LoginMigrationException>>();
+
+                Stopwatch stopWatch = new Stopwatch();
+                stopWatch.Start();
+                exceptionMap.AddExceptions(await loginMigration.ValidateSysAdminPermissions(CancellationToken.None));
+                stopWatch.Stop();
+                TimeSpan elapsedTime = stopWatch.Elapsed;
+
+                LoginMigrationPreValidationResult results = new LoginMigrationPreValidationResult()
+                {
+                    ExceptionMap = exceptionMap,
+                    CompletedStep = LoginMigrationPreValidationStep.SysAdminValidation,
+                    ElapsedTime = MigrationServiceHelper.FormatTimeSpan(elapsedTime)
+                };
+
+                await requestContext.SendResult(results);
+            }
+            catch (Exception e)
+            {
+                await requestContext.SendError(e.ToString());
+            }
+        }
+
+        internal async Task HandleUserMappingValidation(
+            StartLoginMigrationParams parameters,
+            RequestContext<LoginMigrationPreValidationResult> requestContext)
+        {
+            try
+            {
+                ILoginsMigrationLogger logger = this.GetLoginsMigrationLogger();
+                ILoginsMigration loginMigration = new LoginsMigration(parameters.SourceConnectionString, parameters.TargetConnectionString,
+                null, parameters.LoginList, parameters.AADDomainName, logger);
+
+                IDictionary<string, IEnumerable<LoginMigrationException>> exceptionMap = new Dictionary<string, IEnumerable<LoginMigrationException>>();
+
+                Stopwatch stopWatch = new Stopwatch();
+                stopWatch.Start();
+                exceptionMap.AddExceptions(loginMigration.ValidateUserMapping(CancellationToken.None));
```

Comment:
Do we follow any format for keeping all the validations as sync or all as async? As here I can see 2 are async and other 2 are sync.

Reply:
There is no specific format. It varies based on the backend implementation.